### ISSUE TITLE
MqttServer leaks memory (at least UWP)

### ISSUE THREAD

**Issue by tschanko** [chkr1011/MQTTnet#568]
Given the simple scenario of one client subscribing to a certain topic and another client constantly publishing to that topic, memory consumption of the server is constantly increasing. After some profiling it seems that the problem can be located in MqttPendingPacketsQueue and especially in usage of the AsyncAutoResetEvent. I wonder if anyone else encountered this problem?

**Comment by chkr1011** [comment]
Hi,
do you maybe have pending messages in that queue? How fast are your messages published? Once every couple of seconds or a lot more?

I heard from your request a couple of days ago. You can also reach me via e-mail. I am interested in your project and environment around MQTT.

Best regards
Christian

**Comment by tschanko** [comment]
No, the messages are delivered and received in a reasonable amount of time (couple of seconds for 5000 msgs). Using Qos0 all the way. The queue is empty, but passing the CancellationToken all the way into the _queueAutoResetEvent.WaitOneAsync(cancellationToken) leads to the effect that all tasks together with it's resources are kept in memory forever. If you pass CanellationToken.None or leave it out all together the problem goes away and most of the memory gets released.

Rgs,

Thomas

**Comment by chkr1011** [comment]
I guess I found the reason for it. 

https://github.com/chkr1011/MQTTnet/blob/4fed80c2d0c9e08e4364700eaf358bb37ed28d6f/Source/MQTTnet/Internal/AsyncAutoResetEvent.cs#L36

This will create a new task which will delay forever (and requires resources). This should be changed to _Timespan.Zero_ instead. Also the logic can be modified to avoid starting the second task for timeout completely. 

Would you test this solution for me (only replacing the TimeSpan value)? I have to working test setup.

**Comment by tschanko** [comment]
So, I don't need to test it?

**Comment by chkr1011** [comment]
No :-)

**Comment by tschanko** [comment]
Great. Maybe you want to check my fork. I added  a simple UWP App to reproduce the issue. Let me know if I can do something to help you.

**Comment by chkr1011** [comment]
I made a fix for it in the _develop_ branch. Is it possible that you test with the latest _develop_ branch? Version 3.0.0 has some breaking changes but they can be migrated quite fast.

**Comment by tschanko** [comment]
Are you sure that you pushed it? Last commit on develop is 10 days ago

**Comment by chkr1011** [comment]
Yes you are right. I forgot to push. If you don't want to migrate your code to 3.0.0 of this library you can also cherry pick this change at master branch which still has 2.8.x.

**Comment by tschanko** [comment]
I cherrypicked your fix into master and tested it on a WIndows VM. The memory leak is gone. The memory consumption of my UWP is 22MB after startup, 26 MB when the server is started and two clients are connected. When starting the message send loop it raises quickly to around 37 MB and stays there no matter how many messages are processed. I'm going to run the test on an IoT Core board tomorrow. I close the issue now, as I don't expect it to behave any different. 
Thank you for your quick response on this!
Are you planning to do a 2.8.x release including the fix, or should I proceed whith my local build?
Cheers

Thomas
